File: TicTacToe/Private/TicTacToeGridView.m
//
//  TicTacToeGridView.m
//  TicTacToe
//
//  Created by Hélio Mesquita on 23/01/20.
//  Copyright © 2020 Hélio Mesquita. All rights reserved.
//


#import "TicTacToeGridView.h"

@interface TicTacToeGridView ()

@end

@implementation TicTacToeGridView

typedef enum {
    SPACING4,
    SPACING8,
    OFF
} Baseline;


Baseline currentBaseline = OFF;
CGFloat lineSize = 1;
CGFloat lineSpacing = 4;

UIColor * horizontalColor;
UIColor * verticalColor;
NSMutableArray * horizontalBaselines;
NSMutableArray * horizontalSpacing;
NSMutableArray * verticalBaselines;
NSMutableArray * verticalSpacing;

- (instancetype)init
{
    self = [super initWithFrame:UIScreen.mainScreen.bounds];
    if (self) {
        self.translatesAutoresizingMaskIntoConstraints = false;
    }
    return self;
}

- (void)configure {
    [self configureView];
    [self addBaselines];
    [self setBaselineHidden:true];
}

-(void) configureView {
    [self setBackgroundColor: UIColor.clearColor];
    [self setUserInteractionEnabled:false];

    horizontalSpacing = [NSMutableArray array];
    horizontalBaselines = [NSMutableArray array];
    verticalSpacing = [NSMutableArray array];
    verticalBaselines = [NSMutableArray array];
    verticalColor = [UIColor.greenColor colorWithAlphaComponent:0.3];
    horizontalColor = [UIColor.purpleColor colorWithAlphaComponent:0.3];
}

-(void) addBaselines {
    [self startBaselineVertical];
    [self startBaselineHorizontal];
}

- (void)handleTouchEvent {
    switch (currentBaseline) {
        case SPACING4:
            currentBaseline = SPACING8;
            [self setBaselineHidden:false];
            [self setBaselineSpacing:8];
            break;
        case SPACING8:
            currentBaseline = OFF;
            [self setBaselineHidden:true];
            break;
        case OFF:
            currentBaseline = SPACING4;
            [self setBaselineHidden:false];
            [self setBaselineSpacing:4];
            break;
    }
}

-(void) startBaselineVertical {
    int numberOfDivisions = (int)UIScreen.mainScreen.bounds.size.width/lineSpacing;

    UIView * firstBaseline = [self createView: verticalColor];
    [verticalBaselines addObject:firstBaseline];

    [NSLayoutConstraint activateConstraints:@[
        [firstBaseline.topAnchor constraintEqualToAnchor: self.topAnchor],
        [firstBaseline.bottomAnchor constraintEqualToAnchor: self.bottomAnchor],
        [firstBaseline.leadingAnchor constraintEqualToAnchor: self.leadingAnchor],
        [firstBaseline.widthAnchor constraintEqualToConstant: lineSize],
    ]
     ];

    for (int i = 1; i <= numberOfDivisions; i++) {
        UIView * previuosBaseline = verticalBaselines.lastObject;
        UIView * baseline = [self createView:verticalColor];
        NSLayoutConstraint * spacing = [baseline.leadingAnchor constraintEqualToAnchor:previuosBaseline.trailingAnchor constant:lineSpacing];

        [verticalBaselines addObject:baseline];
        [verticalSpacing addObject:spacing];

        [NSLayoutConstraint activateConstraints:@[
            [baseline.topAnchor constraintEqualToAnchor: self.topAnchor],
            [baseline.bottomAnchor constraintEqualToAnchor: self.bottomAnchor],
            [baseline.widthAnchor constraintEqualToConstant: lineSize],
            spacing
        ]
         ];
    }
    [self layoutSubviews];
}

-(void) startBaselineHorizontal {
    int numberOfDivisions = (int)UIScreen.mainScreen.bounds.size.height/lineSpacing;

    UIView * firstBaseline = [self createView: horizontalColor];
    [horizontalBaselines addObject:firstBaseline];

    [NSLayoutConstraint activateConstraints:@[
        [firstBaseline.topAnchor constraintEqualToAnchor: self.topAnchor],
        [firstBaseline.trailingAnchor constraintEqualToAnchor: self.trailingAnchor],
        [firstBaseline.leadingAnchor constraintEqualToAnchor: self.leadingAnchor],
        [firstBaseline.heightAnchor constraintEqualToConstant: lineSize],
    ]
     ];

    for (int i = 1; i <= numberOfDivisions; i++) {
        UIView * previuosBaseline = horizontalBaselines.lastObject;
        UIView * baseline = [self createView:horizontalColor];
        NSLayoutConstraint * spacing = [baseline.topAnchor constraintEqualToAnchor:previuosBaseline.bottomAnchor constant:lineSpacing];

        [horizontalBaselines addObject:baseline];
        [horizontalSpacing addObject:spacing];

        [NSLayoutConstraint activateConstraints:@[
            spacing,
            [baseline.trailingAnchor constraintEqualToAnchor: self.trailingAnchor],
            [baseline.leadingAnchor constraintEqualToAnchor: self.leadingAnchor],
            [baseline.heightAnchor constraintEqualToConstant: lineSize],
        ]
         ];
    }
    [self layoutSubviews];
}

- (void)setBaselineSpacing:(CGFloat)value {
    for (NSLayoutConstraint *constraint in verticalSpacing) {
        [constraint setConstant:value];
    }

    for (NSLayoutConstraint *constraint in horizontalSpacing) {
        [constraint setConstant:value];
    }
}

- (void)setBaselineHidden:(BOOL)hidden {
    for (UIView *subview in verticalBaselines) {
        [subview setHidden:hidden];
    }

    for (UIView *subview in horizontalBaselines) {
        [subview setHidden:hidden];
    }
}

- (nonnull UIView *)createView:(nonnull UIColor *)color {
    UIView *baseline = [[UIView alloc]init];
    baseline.translatesAutoresizingMaskIntoConstraints = false;
    baseline.backgroundColor = color;
    [baseline setUserInteractionEnabled:true];
    [self addSubview:baseline];
    [self sendSubviewToBack:baseline];
    return baseline;
}

@end

File: TicTacToeExample/TicTacToeGridView.swift
//
//  GridView.swift
//  TicTacToeExample
//
//  Created by Hélio Mesquita on 21/01/20.
//  Copyright © 2020 Hélio Mesquita. All rights reserved.
//

import UIKit

class TicTacToeGridView: UIView {

    enum Baseline {
        case SPACING4
        case SPACING8
        case OFF
    }

    let lineSize: CGFloat = 1;
    let lineSpacing: CGFloat = 4;

    var currentBaseline: Baseline = .OFF

    var horizontalBaselines: [UIView] = []
    var horizontalSpacing: [NSLayoutConstraint] = []
    var verticalBaselines: [UIView] = []
    var verticalSpacing: [NSLayoutConstraint] = []

    init() {
        super.init(frame: UIScreen.main.bounds)
        translatesAutoresizingMaskIntoConstraints = false
        configureView()
        addBaselines()
        NotificationCenter.default.addObserver(self, selector: #selector(handleTouchEvent), name: .didChangeBaseline, object: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    func configureView() {
        backgroundColor = UIColor.clear
        isUserInteractionEnabled = false
    }

    func addBaselines() {
        startBaselineVertical()
        startBaselineHorizontal()
//        setBaselineHidden(true)
    }

    @objc func handleTouchEvent() {
        switch (currentBaseline) {
        case .SPACING4:
            currentBaseline = .SPACING8;
            setBaselineHidden(false)
            setBaselineSpacing(8)
            break;
        case .SPACING8:
            currentBaseline = .OFF;
            setBaselineHidden(true)
            break;
        case .OFF:
            currentBaseline = .SPACING4;
            setBaselineHidden(false)
            setBaselineSpacing(4)
            break;
        }
    }

    func startBaselineVertical() {
        let verticalColor = UIColor.green.withAlphaComponent(0.3)
        let numberOfDivisions: Int = Int(UIScreen.main.bounds.width/4)

        let firstBaselineVertical = createView(verticalColor)
        verticalBaselines.append(firstBaselineVertical)
        NSLayoutConstraint.activate([
            firstBaselineVertical.topAnchor.constraint(equalTo: self.topAnchor),
            firstBaselineVertical.bottomAnchor.constraint(equalTo: self.bottomAnchor),
            firstBaselineVertical.leadingAnchor.constraint(equalTo: self.leadingAnchor),
            firstBaselineVertical.widthAnchor.constraint(equalToConstant: lineSize)
        ])

        for _ in 1...numberOfDivisions {
            guard let previuosBaseline = verticalBaselines.last else { continue }
            let baseline = createView(verticalColor)
            let spacing = baseline.leadingAnchor.constraint(equalTo: previuosBaseline.leadingAnchor, constant: lineSpacing)
            verticalBaselines.append(baseline)
            verticalSpacing.append(spacing)

            NSLayoutConstraint.activate([
                baseline.topAnchor.constraint(equalTo: topAnchor),
                baseline.bottomAnchor.constraint(equalTo: bottomAnchor),
                baseline.widthAnchor.constraint(equalToConstant: lineSize),
                spacing
            ])
        }
    }

    func startBaselineHorizontal() {
        let horizontalColor = UIColor.purple.withAlphaComponent(0.3)
        let numberOfDivisions: Int = Int(UIScreen.main.bounds.height/4)

        let firstBaselineHorizontal = createView(horizontalColor)
        horizontalBaselines.append(firstBaselineHorizontal)
        NSLayoutConstraint.activate([
            firstBaselineHorizontal.topAnchor.constraint(equalTo: self.topAnchor),
            firstBaselineHorizontal.leadingAnchor.constraint(equalTo: self.leadingAnchor),
            firstBaselineHorizontal.trailingAnchor.constraint(equalTo: self.trailingAnchor),
            firstBaselineHorizontal.heightAnchor.constraint(equalToConstant: lineSize)
        ])

        for _ in 1...numberOfDivisions {
            guard let previuosBaseline = horizontalBaselines.last else { continue }
            let baseline = createView(horizontalColor)
            let spacing = baseline.topAnchor.constraint(equalTo: previuosBaseline.topAnchor, constant: lineSpacing)
            horizontalBaselines.append(baseline)
            horizontalSpacing.append(spacing)

            NSLayoutConstraint.activate([
                spacing,
                baseline.leadingAnchor.constraint(equalTo: self.leadingAnchor),
                baseline.trailingAnchor.constraint(equalTo: self.trailingAnchor),
                baseline.heightAnchor.constraint(equalToConstant: lineSize)
            ])
        }

    }

    func setBaselineSpacing(_ value: CGFloat) {
        verticalSpacing.forEach { $0.constant = value }
        horizontalSpacing.forEach { $0.constant = value }
    }

    func setBaselineHidden(_ isHidden: Bool) {
        verticalBaselines.forEach { $0.isHidden = isHidden }
        horizontalBaselines.forEach { $0.isHidden = isHidden }
    }

    func createView(_ color: UIColor) -> UIView {
        let baseline = UIView()
        baseline.translatesAutoresizingMaskIntoConstraints = false
        baseline.backgroundColor = color
        baseline.isUserInteractionEnabled = false
        self.addSubview(baseline)
        self.sendSubviewToBack(baseline)
        return baseline
    }

}
